stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+            // The signer is attempting to vote for a round that is greater than one past the last round
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    /// Get this signer's transactions from stackerdb, filtering out any invalid transactions
+    fn get_signer_transactions(
+        &mut self,
+        stacks_client: &StacksClient,
+        current_reward_cycle: u64,
+    ) -> Result<Vec<StacksTransaction>, ClientError> {
+        let transactions: Vec<_> = self
+            .stackerdb
+            .get_current_transactions_with_retry(self.signer_id)?
+            .into_iter()
+            .filter_map(|tx| {
+                self.filter_invalid_transactions(
+                    stacks_client,
+                    current_reward_cycle,
+                    &self.signer_slot_ids,
+                    tx,
+                )
+            })
+            .collect();
+        Ok(transactions)
+    }
+
+    /// Get the transactions that should be included in the block, filtering out any invalid transactions
+    fn get_expected_transactions(
+        &mut self,
+        stacks_client: &StacksClient,
+        current_reward_cycle: u64,
+    ) -> Result<Vec<StacksTransaction>, ClientError> {
+        if self.next_signer_ids.is_empty() {
+            debug!(
+                "Signer #{}: No next signers. Skipping transaction retrieval.",
+                self.signer_id
+            );
+            return Ok(vec![]);
+        }
+        let transactions: Vec<_> = self
+            .stackerdb
+            .get_next_transactions_with_retry(&self.next_signer_ids)?
+            .into_iter()
+            .filter_map(|tx| {
+                self.filter_invalid_transactions(
+                    stacks_client,
+                    current_reward_cycle,
+                    &self.next_signer_slot_ids,
+                    tx,
+                )
+            })
+            .collect();
+        Ok(transactions)
+    }
+
+    /// Determine the vote for a block and update the block info and nonce request accordingly
+    fn determine_vote(
+        signer_id: u32,
+        block_info: &mut BlockInfo,
+        nonce_request: &mut NonceRequest,
+    ) {
+        let mut vote_bytes = block_info.block.header.signer_signature_hash().0.to_vec();
+        // Validate the block contents
+        if !block_info.valid.unwrap_or(false) {
+            // We don't like this block. Update the request to be across its hash with a byte indicating a vote no.
+            debug!(
+                "Signer #{}: Updating the request with a block hash with a vote no.",
+                signer_id
+            );
+            vote_bytes.push(b'n');
+        } else {
+            debug!("Signer #{}: The block passed validation. Update the request with the signature hash.", signer_id);
+        }
+
+        // Cache our vote
+        block_info.vote = Some(vote_bytes.clone());
```

Comment:
Don't we want to be flipping our vote bit here?

Reply:
We do vote no if the block info is valid . I append the 'n' byte only if it fails validation, otherwise it should remain as just the signature hash

Reply:
I have created a helper struct to handle this.